### ISSUE TITLE
OpenSSL 1.1 does not save thread context for threads created by std::thread

### ISSUE THREAD

**Issue by franksinankaya** [openssl/openssl#9361]
This issue has been observed with curl + multi interface using c++ threads. 

There is an implicit assumption that all threads are pthreads on Linux and win32 threads on windows. Modern applications prefer std::thread.

**Comment by richsalz** [comment]
Actually, it's pretty explicit that OpenSSL supports pthreads or win32 threads :)

Does std::thread support per-thread local storage?  Or would you have to subclass std::thread, which would be problematic for app usage.

**Comment by franksinankaya** [comment]
I don't think it is that obvious for majority of the applications.

Yes, thread_local variable attribute can be used.

https://stackoverflow.com/questions/11983875/what-does-the-thread-local-mean-in-c11

**Comment by mspncp** [comment]
> Modern applications prefer std::thread.

I've already used std::thread on linux and my observation was that it's implementation is based on pthreads. So they should coexist in harmony together.

Could you please be more precise about what you mean with "does not save thread context"? 

> This issue has been observed with curl + multi interface using c++ threads.

Do you have any references to issue reports?

**Comment by franksinankaya** [comment]
> Could you please be more precise about what you mean with "does not save thread context"?

Sure, let me give some specifics. 

Issue was observed while building a CURL based http client to cpprestsdk application. OpenSSL works fine with std::thread but the application is leaking memory on each openssl handshake because it tries to store certain values in thread local storage and it fails.

https://github.com/microsoft/cpprestsdk/pull/1182/files#diff-3f0589890efc4b8bbcd9fdaaf6a5fe90R191

As soon as I converted the threads to posix or windows threads, the leak is gone.

I wanted to warn the community about bigger problem as there could be other applications observing the same issue without knowing it.

**Comment by mspncp** [comment]
>  the application is leaking memory on each openssl handshake because it tries to store certain values in thread local storage and it fails.

Is this an interpretation of yours because you are seeing memory leaks, or do you have evidence for your claim? For example, a call stack, the name of the function which failed, the error code....?

Blaming OpenSSL is easy if the memory was allocated by OpenSSL. But it was allocated on behalf of the application and in many cases it turns out that it's the applications fault. I'm not precluding that it might be an issue with OpenSSL. But please give more evidence before "warning the community about big problems".

**Comment by richsalz** [comment]
The fault is with applications that use openssl and don’t look at the openssl docs.  :) That’s not unique, projects often don’t know details of their dependencies. :(

The warning is worthwhile, but it’s this: *OpenSSL currently only supports pthreads and native Win32 threads.*

In the link you posted, it seems there is a new C++ type for thread-local storage.  Using that would require rewriting some low-level OpenSSL internals which currently use the native thread-local storage. Doing it without being overly invasive or forcing most of OpenSSL to be built with C++ will be tricky, particularly around the cleanup when the thread is terminated.

**Comment by mspncp** [comment]
@franksinankaya we would appreciate it very much if you could provide us with a small compilable example program which exhibits the issue. If we can reproduce the problem, we might be able to fix it.

**Comment by franksinankaya** [comment]
```
/***************************************************************************
 *                                  _   _ ____  _
 *  Project                     ___| | | |  _ \| |
 *                             / __| | | | |_) | |
 *                            | (__| |_| |  _ <| |___
 *                             \___|\___/|_| \_\_____|
 *
 * Copyright (C) 1998 - 2018, Daniel Stenberg, <daniel@haxx.se>, et al.
 *
 * This software is licensed as described in the file COPYING, which
 * you should have received as part of this distribution. The terms
 * are also available at https://curl.haxx.se/docs/copyright.html.
 *
 * You may opt to use, copy, modify, merge, publish, distribute and/or sell
 * copies of the Software, and permit persons to whom the Software is
 * furnished to do so, under the terms of the COPYING file.
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
 * KIND, either express or implied.
 *
 ***************************************************************************/
/* <DESC>
 * A basic application source code using the multi interface doing two
 * transfers in parallel.
 * </DESC>
 */
 
#include <thread>
#include <mutex>
#include <atomic>

#include <stdio.h>
#include <string.h>

/* somewhat unix-specific */
#include <sys/time.h>
#include <unistd.h>

/* curl stuff */
#include <curl/curl.h>

/*
 * Download a HTTP file and upload an FTP file simultaneously.
 */

#define HANDLECOUNT 2   /* Number of simultaneous transfers */
CURL *handles[HANDLECOUNT];
CURLM *multi_handle;
std::mutex mtx;
std::atomic<int> transfercount;

void commfunc(void)
{
  CURLMsg *msg; /* for picking up messages with the transfer status */
  int msgs_left; /* how many messages are left */
  int still_running = 0; /* keep number of running handles */

  /* we start some action by calling perform right away */
  mtx.lock();
  curl_multi_perform(multi_handle, &still_running);
  mtx.unlock();

  while(still_running) {
    struct timeval timeout;
    int rc; /* select() return code */
    CURLMcode mc; /* curl_multi_fdset() return code */

    fd_set fdread;
    fd_set fdwrite;
    fd_set fdexcep;
    int maxfd = -1;

    long curl_timeo = -1;

    FD_ZERO(&fdread);
    FD_ZERO(&fdwrite);
    FD_ZERO(&fdexcep);

    /* set a suitable timeout to play around with */
    timeout.tv_sec = 1;
    timeout.tv_usec = 0;

    mtx.lock();
    curl_multi_timeout(multi_handle, &curl_timeo);
    mtx.unlock();
    if(curl_timeo >= 0) {
      timeout.tv_sec = curl_timeo / 1000;
      if(timeout.tv_sec > 1)
        timeout.tv_sec = 1;
      else
        timeout.tv_usec = (curl_timeo % 1000) * 1000;
    }

    /* get file descriptors from the transfers */
    mtx.lock();
    mc = curl_multi_fdset(multi_handle, &fdread, &fdwrite, &fdexcep, &maxfd);
    mtx.unlock();

    if(mc != CURLM_OK) {
      fprintf(stderr, "curl_multi_fdset() failed, code %d.\n", mc);
      break;
    }

    /* On success the value of maxfd is guaranteed to be >= -1. We call
       select(maxfd + 1, ...); specially in case of (maxfd == -1) there are
       no fds ready yet so we call select(0, ...) --or Sleep() on Windows--
       to sleep 100ms, which is the minimum suggested value in the
       curl_multi_fdset() doc. */

    if(maxfd == -1) {
#ifdef _WIN32
      Sleep(100);
      rc = 0;
#else
      /* Portable sleep for platforms other than Windows. */
      struct timeval wait = { 0, 100 * 1000 }; /* 100ms */
      rc = select(0, NULL, NULL, NULL, &wait);
#endif
    }
    else {
      /* Note that on some platforms 'timeout' may be modified by select().
         If you need access to the original value save a copy beforehand. */
      rc = select(maxfd + 1, &fdread, &fdwrite, &fdexcep, &timeout);
    }

    switch(rc) {
    case -1:
      /* select error */
      break;
    case 0: /* timeout */
    default: /* action */
      mtx.lock();
      curl_multi_perform(multi_handle, &still_running);
      mtx.unlock();
      break;
    }
  }

  /* See how the transfers went */
  while(1) {
    mtx.lock();
    msg = curl_multi_info_read(multi_handle, &msgs_left);
    mtx.unlock();
    if (!msg)
        break;
    if(msg->msg == CURLMSG_DONE) {
      printf("HTTP transfer %p completed with status %d\n", msg->easy_handle, msg->data.result);
      transfercount++;
      curl_easy_cleanup(msg->easy_handle);      
    }
  }    
}

void commthread(void)
{
    while (1)
        commfunc();
}

int RequestCount = 100;

void requestthread(void)
{
    for (int i =0; i < RequestCount; i++)
    {
        CURL *handle = curl_easy_init();
        curl_easy_setopt(handle, CURLOPT_URL, "https://example.com");
        mtx.lock();
        curl_multi_add_handle(multi_handle, handle);
        mtx.unlock();
    }
    printf("%d requests queued\n", RequestCount);
}

std::thread commthr;
std::thread requestthr;

int main(void)
{

  int i;

  transfercount = 0;

  /* Allocate one CURL handle per transfer */
  for(i = 0; i<HANDLECOUNT; i++)
  {
    handles[i] = curl_easy_init();
    curl_easy_setopt(handles[i], CURLOPT_URL, "https://www.google.com");
  }

  /* init a multi stack */
  multi_handle = curl_multi_init();

  /* add the individual transfers */
  for(i = 0; i<HANDLECOUNT; i++)
  {
    mtx.lock();
    curl_multi_add_handle(multi_handle, handles[i]);
    mtx.unlock();
  }

  commthr = std::thread(commthread);
  commthr.detach();
  requestthr = std::thread(requestthread);
  requestthr.join();
  while (transfercount != (2 + RequestCount));
  

  return 0;
}

```

g++ multi-app.cpp -lcurl -lpthread -o multi-app

**Comment by franksinankaya** [comment]
heaptrack ./multi-app

**Comment by franksinankaya** [comment]
[heaptrack.multi-app.16850.gz](https://github.com/openssl/openssl/files/3390298/heaptrack.multi-app.16850.gz)

**Comment by franksinankaya** [comment]
![2019-07-14 14_18_09-Heaptrack - heaptrack multi-app 16850 gz](https://user-images.githubusercontent.com/41809318/61187659-48be6380-a642-11e9-84d9-9fecaceba25a.png)

**Comment by franksinankaya** [comment]
![2019-07-14 14_19_15-Heaptrack - heaptrack multi-app 16850 gz](https://user-images.githubusercontent.com/41809318/61187667-64296e80-a642-11e9-8b3f-0d218533f963.png)

**Comment by franksinankaya** [comment]
[heaptrack.multi-app.17129.gz](https://github.com/openssl/openssl/files/3390307/heaptrack.multi-app.17129.gz)

**Comment by mspncp** [comment]
Thank you very much @franksinankaya for going through the effort to provide us with an example. It might take a little while, but we are going to take a look at it.

**Comment by mspncp** [comment]
According to your log it seems like you are using the openssl library provided by your linux distribution, correct?

```
s /usr/lib/x86_64-linux-gnu/libssl.so.1.1
s /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
```

From the so suffix it is not clear whether it's an OpenSSL version 1.1.0 or 1.1.1. Could you please tell us which linux distribution you are using and post the output of the `openssl version` command? Thanks in advance.

**Comment by franksinankaya** [comment]
That's correct. The snapshot was taken from ubuntu 18.04 WSL.

openssl version
OpenSSL 1.1.0g  2 Nov 2017

We even checked out latest OpenSSL 1.1.1 branch. The behavior didn't change.

**Comment by richsalz** [comment]
I am confused why there is still discussion.  OpenSSL does not support ```std::thread```  Look at what the sample code does with ```commthr``` -- it creates a ```std::thread```, *detaches it*, does some curl stuff which presumably does some openssl stuff, and keeps going until process exits. (I would assume the underlying pthread-join is done, and the thread-cleanup is done at that point.)

The fact that, *as the original poster said,* "As soon as I converted the threads to posix or windows threads, the leak is gone." should be a big red flag.  _OpenSSL does not support std::thread_

What am I missing?

**Comment by franksinankaya** [comment]
That's where the problem is..

OpenSSL 1.0 allowed users to declare their own locks so that OpenSSL would work regardless of what thread implementation a particular OS/app used.

Now, OpenSSL 1.1 thinks that it is smarter and does the TLS stuff on behalf of the user with a certain thread model assumptions and stubbed out all lock functions.

While doing this transition, OpenSSL 1.1 threw C++ users under the bus.

If OpenSSL developers acknowledge "no C++ threads support explicitly" somewhere or with this issue rather than hoping that pthread would save the world somewhere, I'm OK closing the issue.

I spent a week debugging this and was very surprised to find out this issue. I sure don't wish anybody this debugging experience.

**Comment by mspncp** [comment]
> What am I missing?

Curiosity?  ;-)

But now seriously: Although OpenSSL has remained with the C90 standard in all these years for maximum portability, it is equally true that OpenSSL served as crypto library for many programs written in modern C and C++ languages, and many others. This includes software from my company, too.

Now after all these years of platforms specific thread API ~~confusion~~ diversity, finally true language support for threads arrived in C++11. Which is beneficent for the C++ community. In my company we started switching to std::thread in newer software, and I like the new interface. It is inspired by and close to what pthreads provides. My experience up to now (gcc toolchain on linux) is that the std::thread class is a thin wrapper around the pthreads library. So in principle I don't see any contradiction with your claim that OpenSSL only support pthreads. Maybe it's just a little problem which can be fixed, and I'm curious to find it out.

**Comment by mspncp** [comment]
But @richsalz mentioned an important observation:  you are detaching your `commthr` and then you exit without joining it. Then you exit main with a thread still running. Could this be the cause for your memory leak? Does your test program also leak when you join both threads?

**Comment by franksinankaya** [comment]
I tried to create a mock application rather than giving the full application. ( I can get the real application if we can't reproduce. It is not a very big deal.)

I'm waiting for all read requests to be serviced with a counter before leaving the application. We know for sure there is no outstanding requests in flight. The leak is happening on each send request not during thread termination.

If you increase the number of requests also space them with sleep, and then watch the memory usage in parallel, it climbs all the way up to 4GB before application getting terminated. Runtime memory leak is pointing to crypto_zalloc function.

If I look at the memory leaks at the end of application exit, there are still some leaks but only few.

**Comment by franksinankaya** [comment]
Another possibility is to compare the number of allocations compared with std::thread and pthread using the same test application above. This might be easier.

**Comment by mspncp** [comment]
> @bernd-edlinger bernd-edlinger unpinned this issue 6 hours ago

Oh, sorry. I wasn't aware that it's pinned for everybody. That makes pinning a pretty useless GitHub feature.

**Comment by bernd-edlinger** [comment]
haha...
I definitely thought I close a pop-up window.

**Comment by mspncp** [comment]
> If you increase the number of requests also space them with sleep, and then watch the memory usage in parallel, it climbs all the way up to 4GB before application getting terminated. Runtime memory leak is pointing to crypto_zalloc function.
> 
> If I look at the memory leaks at the end of application exit, there are still some leaks but only few.


@franksinankaya I can confirm that your example shows the memory usage pattern you describe: The heaptrack plot shows linear growth until shortly before the end, when the usage drops back to almost zero. And I have a simple explanation for this behaviour of you application.

_BTW, thanks for introducing me to heaptrack, it's a very neat tool. And it's fast. Initially I tried valgrind with your example, but that was hopelessly slow. But in the end, it was neither valgrind nor heaptrack which provided the answer, it was gdb._

**And apologies in advance for the joke, but I couldn't resist: the error lies in layer 7 or 8, not in layer 4.** 

In the attached [gdb.multi-app.log.gz](https://github.com/openssl/openssl/files/3398817/gdb.multi-app.log.gz) you find the answer: your application calls `curl_easy_init()` 103 times in a row, creating 103 parallel sessions, before finally releasing 102 of the sessions with `curl_easy_cleanup()`, leaking one session.

```
msp@msppc:~/openssl/issues/9361-std-thread$ grep -n 'Breakpoint.*curl_easy' gdb.multi-app.log

...
25:Thread 3 "multi-app" hit Breakpoint 1, curl_easy_init () at easy.c:293
29:Thread 3 "multi-app" hit Breakpoint 1, curl_easy_init () at easy.c:293
33:Thread 3 "multi-app" hit Breakpoint 1, curl_easy_init () at easy.c:293
...
415:Thread 3 "multi-app" hit Breakpoint 1, curl_easy_init () at easy.c:293
419:Thread 3 "multi-app" hit Breakpoint 1, curl_easy_init () at easy.c:293
423:Thread 3 "multi-app" hit Breakpoint 1, curl_easy_init () at easy.c:293
629:Thread 2 "multi-app" hit Breakpoint 2, curl_easy_cleanup (data=0x5555555a7988) at easy.c:757
633:Thread 2 "multi-app" hit Breakpoint 2, curl_easy_cleanup (data=0x5555555a2218) at easy.c:757
637:Thread 2 "multi-app" hit Breakpoint 2, curl_easy_cleanup (data=0x7fffdc00be08) at easy.c:757
...
1025:Thread 2 "multi-app" hit Breakpoint 2, curl_easy_cleanup (data=0x7fffdc22a178) at easy.c:757
1029:Thread 2 "multi-app" hit Breakpoint 2, curl_easy_cleanup (data=0x7fffdc224808) at easy.c:757
1033:Thread 2 "multi-app" hit Breakpoint 2, curl_easy_cleanup (data=0x7fffdc000b28) at easy.c:757

msp@msppc:~/openssl/issues/9361-std-thread$ grep -n 'Breakpoint.*curl_easy_init' gdb.multi-app.log | wc -l
103

msp@msppc:~/openssl/issues/9361-std-thread$ grep -n 'Breakpoint.*curl_easy_cleanup' gdb.multi-app.log | wc -l
102
```

I did not go into the details of your example, I only compared it with the original curl [multi-app.c](https://github.com/curl/curl/blob/master/docs/examples/multi-app.c) example and noticed that you made some nontrivial changes. I looks like me that you are firing off all connections simultaneously and not trying to multiplex them over a limited number of channels.

Hopefully, this information will help you to fix the bug in your http client.

**Comment by mspncp** [comment]
As for the original question whether OpenSSL works with std::thread's:  the answer is clearly yes if std::thread is implemented on top of the pthread library.  I verified it with gcc on linux, but I'm confident that similar results hold on other posix compliant operating systems with pthreads available and also for other compilers (clang). And I would be very much surprised if Microsoft's std::thread would'nt coexist equally nice with OpenSSL on Windows, because it is designed to cooperate with the native Win32 thread api.

 I'll post a writeup soon about my results.

**Comment by mspncp** [comment]
> Hopefully, this information will help you to fix the bug in your http client.

I recommend to build not only libcrypto and libssl, but also libcurl with debug information,  then you'll see the complete stack and it's easier to understand what's going on.

**Comment by franksinankaya** [comment]
How do you explain the linear growth that does not happen if you use pthread instead of std::thread?

As I said, I'm not worried about memory leaks in shutdown. Runtime memory consumption is the main concern here.

**Comment by mspncp** [comment]
> How do you explain the linear growth that does not happen if you use pthread instead of std::thread?

My guess: you changed more in your program than just converting from pthread to std::thread.

To switch from a pthread to std::thread and vice versa, only roughly 4 lines need to be changed: 1) the inlude directive, 2) the function signature of the thread function, 3) the thread creation call, and 4) the thread join call. My bet is: if you only do these simple changes for both of your std:thread's, then the resulting pthreads program will exhibit exactly the same behaviour. 

You can prove me wrong by postiing a counterexample (std::thread vs. pthread implementation). But be warned, I'll send it through diff first :-)

**Comment by levitte** [comment]
Can we stop the `std::thread` vs `pthread` rethorics, please?  `std::thread` is a class layer on top of `pthread` on Linux, so putting them as alternatives of each other *serves no purpose and is a red herring*.

What you're doing when you're using `pthread` directly is that you're skipping a layer.  Simple as that.  And if the difference in behaviour is exhibited by the presence of the extra layer, I would start looking at that extra layer and what's going on there.

**Comment by mspncp** [comment]
> Can we stop the std::thread vs pthread rethorics, please? std::thread is a class layer on top of pthread on Linux, so putting them as alternatives of each other serves no purpose and is a red herring.

@levitte I made extensive tests in the last days and I am currently working on a writeup to bust the myth that OpenSSL does not work with std:thread, only with the traditional pthreads resp. Win32 threads. I'll post it as a separate issue, so it can be more easily found and/or referenced.

**Comment by mspncp** [comment]
> And if the difference in behaviour is exhibited by the presence of the extra layer, I would start looking at that extra layer and what's going on there.

As for the example above, I already [proved](https://github.com/openssl/openssl/issues/9361#issuecomment-511963604) that the error lies in the application code.

**Comment by mspncp** [comment]
Closing this issue, as it is not an OpenSSL bug.

**Comment by mspncp** [comment]
@ChrisZhangJin it is **not true** that OpenSSL can't be used with std::thread: The std::thread template class is just a thin wrapper on top of the pthread library. Since OpenSSL supports pthreads, it is no problem if a C++ application uses std::thread and links to the OpenSSL library.

Please read the discussion more closely, in particular https://github.com/openssl/openssl/issues/9361#issuecomment-511967417 and https://github.com/openssl/openssl/issues/9361#issuecomment-512176326.

**Comment by mspncp** [comment]
If anyone thinks there is an issue with OpenSSL and std::thread's, please file an issue and (ideally) provide a small C++ program to reproduce the issue. Otherwise, please stop spreading fake news ;-).

**Comment by ChrisZhangJin** [comment]
> @ChrisZhangJin it is **not true** that OpenSSL can't be used with std::thread The std::thread template class is just a thin wrapper on top of the pthread library. Since OpenSSL supports pthreads, it is no problem if a C++ application uses std::thread and links to the OpenSSL library.
> 
> Please read the discussion more closely, in particular [#9361 (comment)](https://github.com/openssl/openssl/issues/9361#issuecomment-511967417) and [#9361 (comment)](https://github.com/openssl/openssl/issues/9361#issuecomment-512176326).
-------------------------------------------------------------------
Sorry for my misleading words.   I'm not meaning "not supported" or "cannot be used".  Actually it's working pretty fine. Anyway, i encountered the same "thing" with @franksinankaya did.  when i coding it with std::thread, the momery leaks. And I changed the code with pthread, suprisingly the memory leaks are going.  

That's what i found.   No offensive to the effort of contributors' great work.  No intent to spead the fakes.
no more questions. Thanks for your reply.  :)

**Comment by mspncp** [comment]
@ChrisZhangJin sorry, you didn't tell us that you also encountered a similar problem when using OpenSSL with std::thread. Would it be possible to provide a small compilable example with which we can reproduce your problem? If yes, please create a separate issue (you can refer to this issue by mentioning #9361) for it and I will take a look.

**Comment by mspncp** [comment]
Ideally, it would be great if you could provide both examples,  using pthread vs. using std::thread.
